crates/services/p2p/src/codecs/request_response.rs

```diff
@@ -0,0 +1,153 @@
+    // TODO: https://github.com/FuelLabs/fuel-core/issues/2459.
+    // Make this a u64 instead of usize.
+    pub(crate) max_response_size: usize,
+}
+
+/// Since Postcard does not support async reads or writes out of the box
+/// We prefix Request & Response Messages with the length of the data in bytes
+/// We expect the substream to be properly closed when response channel is dropped.
+/// Since the request protocol used here expects a response, the sender considers this
+/// early close as a protocol violation which results in the connection being closed.
+/// If the substream was not properly closed when dropped, the sender would instead
+/// run into a timeout waiting for the response.
+#[async_trait]
+impl<Codec> request_response::Codec for RequestResponseMessageHandler<Codec>
+where
+    Codec: Encode<RequestMessage, Error = io::Error>
+        + Decode<RequestMessage, Error = io::Error>
+        + Encode<V1ResponseMessage, Error = io::Error>
+        + Decode<V1ResponseMessage, Error = io::Error>
+        + Encode<V2ResponseMessage, Error = io::Error>
+        + Decode<V2ResponseMessage, Error = io::Error>
+        + Send,
+{
+    type Protocol = RequestResponseProtocol;
+    type Request = RequestMessage;
+    type Response = V2ResponseMessage;
+
+    async fn read_request<T>(
+        &mut self,
+        _protocol: &Self::Protocol,
+        socket: &mut T,
+    ) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let mut response = Vec::new();
+        socket
+            .take(self.max_response_size as u64)
+            .read_to_end(&mut response)
+            .await?;
+        self.codec.decode(&response)
+    }
+
+    async fn read_response<T>(
+        &mut self,
+        protocol: &Self::Protocol,
+        socket: &mut T,
+    ) -> io::Result<Self::Response>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let mut response = Vec::new();
+        socket
+            .take(self.max_response_size as u64)
+            .read_to_end(&mut response)
+            .await?;
+
+        match protocol {
+            RequestResponseProtocol::V1 => {
+                let v1_response: V1ResponseMessage = self.codec.decode(&response)?;
+                Ok(v1_response.into())
+            }
+            RequestResponseProtocol::V2 => self.codec.decode(&response),
+        }
+    }
+
+    async fn write_request<T>(
+        &mut self,
+        _protocol: &Self::Protocol,
+        socket: &mut T,
+        req: Self::Request,
+    ) -> io::Result<()>
+    where
+        T: futures::AsyncWrite + Unpin + Send,
+    {
+        let encoded_data = self.codec.encode(&req)?;
+        socket.write_all(&encoded_data.as_bytes()).await?;
+        Ok(())
+    }
+
+    async fn write_response<T>(
+        &mut self,
+        protocol: &Self::Protocol,
+        socket: &mut T,
+        res: Self::Response,
+    ) -> io::Result<()>
+    where
+        T: futures::AsyncWrite + Unpin + Send,
```

Comment:
Nit, for consistency with the `AsyncRead` above.

```suggestion
        T: AsyncWrite + Unpin + Send,
```

Reply:
refactored in 5063decfe6